flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureRateRestartBackoffTimeStrategy.java

```diff
@@ -0,0 +1,156 @@
+
+	FailureRateRestartBackoffTimeStrategy(Clock clock, int maxFailuresPerInterval, long failuresIntervalMS, long backoffTimeMS) {
+
+		checkArgument(maxFailuresPerInterval > 0, "Maximum number of restart attempts per time unit must be greater than 0.");
+		checkArgument(failuresIntervalMS > 0, "Failures interval must be greater than 0 ms.");
+		checkArgument(backoffTimeMS >= 0, "Backoff time must be at least 0 ms.");
+
+		this.failuresIntervalMS = failuresIntervalMS;
+		this.backoffTimeMS = backoffTimeMS;
+		this.maxFailuresPerInterval = maxFailuresPerInterval;
+		this.failureTimestamps = new ArrayDeque<>(maxFailuresPerInterval);
+		this.strategyString = generateStrategyString();
+		this.clock = checkNotNull(clock);
+	}
+
+	@Override
+	public boolean canRestart() {
+		if (isFailureTimestampsQueueFull()) {
+			Long now = clock.absoluteTimeMillis();
+			Long earliestFailure = failureTimestamps.peek();
+
+			return (now - earliestFailure) > failuresIntervalMS;
+		} else {
+			return true;
+		}
+	}
+
+	@Override
+	public long getBackoffTime() {
+		return backoffTimeMS;
+	}
+
+	@Override
+	public void notifyFailure(Throwable cause) {
+		if (isFailureTimestampsQueueFull()) {
+			failureTimestamps.remove();
+		}
+		failureTimestamps.add(clock.absoluteTimeMillis());
+	}
+
+	@Override
+	public String toString() {
+		return strategyString;
+	}
+
+	private boolean isFailureTimestampsQueueFull() {
+		return failureTimestamps.size() >= maxFailuresPerInterval;
+	}
+
+	private String generateStrategyString() {
+		StringBuilder str = new StringBuilder("FailureRateRestartBackoffTimeStrategy(");
+		str.append("FailureRateRestartBackoffTimeStrategy(failuresIntervalMS=");
+		str.append(failuresIntervalMS);
+		str.append(",backoffTimeMS=");
+		str.append(backoffTimeMS);
+		str.append(",maxFailuresPerInterval=");
+		str.append(maxFailuresPerInterval);
+		str.append(")");
+
+		return str.toString();
+	}
+
+	public static FailureRateRestartBackoffTimeStrategyFactory createFactory(final Configuration configuration) {
+		return new FailureRateRestartBackoffTimeStrategyFactory(configuration);
+	}
+
+	/**
+	 * The factory for creating {@link FailureRateRestartBackoffTimeStrategy}.
+	 */
+	public static class FailureRateRestartBackoffTimeStrategyFactory implements RestartBackoffTimeStrategy.Factory {
+
+		private final int maxFailuresPerInterval;
+
+		private final long failuresIntervalMS;
+
+		private final long backoffTimeMS;
+
+		public FailureRateRestartBackoffTimeStrategyFactory(final Configuration configuration) {
+			this.maxFailuresPerInterval = configuration.getInteger(
+				RESTART_BACKOFF_TIME_STRATEGY_FAILURE_RATE_MAX_FAILURES_PER_INTERVAL);
+			this.failuresIntervalMS = getIntervalMSFromConfiguration(
+				configuration,
+				RESTART_BACKOFF_TIME_STRATEGY_FAILURE_RATE_FAILURE_RATE_INTERVAL);
+			this.backoffTimeMS = getIntervalMSFromConfiguration(
+				configuration,
+				RESTART_BACKOFF_TIME_STRATEGY_FAILURE_RATE_FAILURE_RATE_BACKOFF_TIME);
```

Comment:
I would move the configuration parsing logic into the `createFactory` method because it now mixes concerns.

Reply:
OK